### DISCUSSION TITLE
Some bugs in amiberry 3.4 64Bit

### DISCUSSION THREAD

**Discussion by Retro1968** [BlitterStudio/amiberry#769] [Q&A] [answered]
System: Raspberry Pi 4, Manjaro Linux, amiberry-dev 3.4 beta  (22.01.2020) 64 Bit.
AmigaOS 3.1.4.1 using host-run and Rabbitholes

Hello midwan,
I would like to report a few bugs.

1. In version 3.4 my Rabbitholes no longer work.
A click on the link icon (e.g. Firefox) leads to an immediate crash of amiberry. This works fine under amiberry 3.3.

2. The function "Display -> Blacker than black" behaves in exactly the opposite way. If the function is deactivated you get a real black. If the function is activated you only get a pale black.

3. The mouse pointer jerks regularly. Not always but every now and then. I had previously updated Manjaro Linux and recompiled amiberry 3.4.

Cheers
Retro1968

P.S .: Thanks for your great work, buddy.

**Comment by Retro1968** [comment]
The mouse jerking has disappeared with the version dated January 23, 2021 (JIT correction).

**Comment by midwan** [comment]
For item 1, please make sure you're using the latest version of `host-run` as well.

**Comment by Retro1968** [comment]
Do you have a download link?

**Comment by midwan** [comment]
https://github.com/midwan/host-run/releases

**Comment by midwan** [comment]
As for 2, it behaves exactly as it does in WinUAE. Perhaps this is an upstream bug?

**Comment by midwan** [comment]
Reported to Toni: https://github.com/tonioni/WinUAE/issues/189

**Comment by Retro1968** [comment]
Amiberry also crashes with the new host-run V1.3.

**Comment by midwan** [comment]
It seems to be working for me. Have you enabled "Allow native code" in the Misc panel?
How are you calling host-run, what parameters are you using?

**Comment by Retro1968** [comment]
"Allow native code" is activated.

host-run filezilla

And mouse-jerking is back.

**Comment by Retro1968** [comment]
Amiberry also crashes when I enter "host-run filezilla" in the Amiga shell.

**Comment by midwan** [comment]
I've just re-tested it under AmiKit and a vanilla 3.9 installation, both worked normally for me.
But under 3.1.4.1 it closes down the emulator. I'll check why...

**Comment by midwan** [comment]
Actually, it worked after I recreated my old 3.1.4 config - might have been some invalid option in the .uae file.
Can you try that? Keep a copy of your old one if you want, but create a new one using the latest beta GUI.

**Comment by midwan** [comment]
Regarding the gfx_blackerthanblack option, Toni replied:
"It is correct, it slightly increases brightness of "normal" black to make it look like borderblanked black is blacker. It is impossible (at least without HDR) to make actual blacker black."

**Comment by Retro1968** [comment]
I have created a new config file with the old values. Amiberry still crashes when I click a Rabbithole icon or enter "host-run firefox" in the amiga shell.

I get a blacker black (a real black black) if the option is disabled as i wrote above.

"The function "Display -> Blacker than black" behaves in exactly the opposite way. If the function is deactivated you get a real black. If the function is activated you only get a pale black."  Test it yourself with R-Type 2.

**Comment by midwan** [comment]
OK, please enable logging and send me the amiberry logfile that gets created, when you try to run host-run.

Regarding the blacker than black option, I'm not sure you understood what Toni said. That's by design, it increases the RGB color brightness by 10, so that it would look like the blanked border is a darker shade.

**Comment by Retro1968** [comment]
It is really annoying when the mouse pointer stops and jerks every now and then when moving the mouse.

OK, so the option "Blacker than black" has to work like this ... ;-)

Do I have to activate "Log illegal memory access" for the log file or how does it work?

**Comment by midwan** [comment]
Go to the Paths panel, and enable Logging. The path it's saved in, is in the same directory it started from by default. But you can change it, of course

**Comment by Retro1968** [comment]
Here we go...
[amiberry.log](https://github.com/midwan/amiberry/files/5860345/amiberry.log)

**Comment by midwan** [comment]
Is that the complete logfile? It seems to be missing lots of lines, before the crash... Did you edit it?

**Comment by Retro1968** [comment]
I have activated "Enabled loggin" and saved it in the config. After that, I ended amiberry. After I restarted amiberry, "Enabled logging" was deactivated again. Now I have activated it again and entered "host-run firefox" in the amiga shell. After that, amiberry crashed. You now get the log file. It was NOT edited by me just like before.

Please say something about the fact that I have these severe mouse lags. When I entered "host-run firefox" into the amiga shell, there were lags, so I had to wait until I could continue typing. That is a big problem!!!

[amiberry.log](https://github.com/midwan/amiberry/files/5860885/amiberry.log)

**Comment by midwan** [comment]
Hm, this one is also missing parts of the initialization phase, which might help understand what's happening.
Please try the following:
- Enable logging in Paths.
- Quit the emulator
- Start it up again, then proceed to boot your environment and run `host-run firefox` for example
- Get the new logfile and attach it here

Hopefully that should give us some more details.

Regarding mouse lags, I'm not sure what you mean. I don't see anything like that here, nor do any of the other testers we're working with these last months, getting the beta bugfixed? Can you record a video to demonstrate please?

**Comment by Retro1968** [comment]
That's exactly what I tried, but "Enable Logging" is NOT saved in the config, but is always deactivated on restart, as I described above.

**Comment by Retro1968** [comment]
amiberry always takes short breaks of 1-2 seconds. Then the mouse stops or if you try to enter something in the amiga shell.

**Comment by midwan** [comment]
@Retro1968 
It should always save those settings when you Quit - if it doesn't, then perhaps you have a permissions problem in your `conf/amiberry.conf` file (or the whole conf directory?), since this option is a global one and gets saved there.

**Comment by Retro1968** [comment]
I completely reinstalled Manjaro Linux and amiberry. All the problems described above, including the mouse and keyboard lag, are still there with amiberry V3.4. But even with amiberry V3.3, the mouse and keyboard stutter. Mouse and keyboard entries are repeatedly not possible for about 1-2 seconds. Maybe a Manjaro problem. I once opened a thread here.

https://forum.manjaro.org/t/problems-running-amiberry-since-a-manjaro-update/50370


Neither V3.3 nor V3.4 save the "Enable Logging" option, although I had previously executed these commands:

sudo chown -R hybrid:hybrid /home/hybrid
chmod -R 755 /home/hybrid

**Comment by midwan** [comment]
I found that if you run `host-run` and you didn't have the Virtual Mouse Driver enabled, Amiberry would crash.
That's fixed with 4dc31273f0d78461bd66c070a13b7f26d6f147b2

Nobody else seems to be able to recreate the mouse/keyboard stutter you describe however, so I doubt that's an Amiberry issue. Nor a Manjaro issue specifically, as we have several people testing there at the moment. Have you tried fully upgrading the system? Otherwise, maybe something else (faulty SDcard?) is affect the system.

**Comment by Retro1968** [comment]
Hello Midwan,
I used a different SD card. Using the attached screenshot, you can see that the SD card was previously tested with the "H2testw 1.4" program. The result is in german "Fertig, kein Fehler aufgetreten.", in english ("Done, no error occurred.").

![2021 01 24 - SDTest](https://user-images.githubusercontent.com/58906142/105696722-fc750480-5f03-11eb-8188-63e8395f0890.jpg)

With the new amiberry 3.4 version, the Rabbitholes work again. Thanks!

The jerking problem remains! I have a guess and ask you to investigate. So that the Rabbitholes worked again, you fixed a problem with the MOUSE DRIVER. I'll also report problems with the MOUSE ... the short dropouts, the mouse jerks. Isn't that an indication that there is still a mouse driver problem or something similar somewhere?

Cheers
Retro1968

**Comment by Retro1968** [comment]
So in the end the DPI thing wasn't confirmed. The mouse works perfectly on other computers and also on the RPi 4 with RetroPie and amiberry 3.3 32-bit. Until recently, this was also possible under amiberry 64Bit. Maybe a driver problem after all?

I don't have the problem with another mouse. Sure that I am currently using that other mouse on the RPi4. But why does the first mouse no longer run correctly only under amiberry 64Bit?!?

**Comment by solskogen** [comment]
You could try to add usbhid.mousepoll=0 to /boot/cmdline.txt

**Comment by Retro1968** [comment]
> You could try to add usbhid.mousepoll=0 to /boot/cmdline.txt

Thank you for your help. Here is the result:
The mouse could be used normally under the Linux interface. Under amiberry it was much worse. The mouse behaved chaotically and was no longer usable at all.

**Comment by Retro1968** [comment]
So guys, I'll let you know the status quo.

I have now tried various other mice, another 64-bit Linux distribution (Ubuntu), another 32-bit Linux distribution (Rasperry Pi OS), amiberry 3.3 + 4 beta and I keep coming back to the result that there are no problems under 32 bit , but there are ALWAYS problems under 64 bit. I always have short dropouts in which amiberry take breaks. But even with Linux 64 bit itself I sometimes have these problems.

Assumption: Could it be that some process is regularly writing to the SD card and causing the jerking? Reading processes do not seem to cause any problems. I doubt the SD card is causing any problems as this is already the second SD card I'm trying and both SD cards have been pre-tested and are error free and work under 32-Bit without problems.

When I installed Manjaro Linux 64 bit with AmiHybrid in September 2020 everything ran without errors. I am annoyed with myself that I no longer have a backup copy of it ... 

I installed Rasberry Pi OS with AmiHybrid during my tests. The only stupid thing is that the rabbitholes didn't worked. The programs were started in the background, but I couldn't bring them into the foreground. I ran amiberry in fullscreen, windowed and full-windowed. Does anyone know a solution here or do the Rabbiholdes don't work under 32 bit? 

Cheers
SID-6581

**Comment by Retro1968** [comment]
I found another problem that occurs in every version of amiberry. 

The current P96 driver from October 2020 causes problems. For example, if you open Deluxe Paint V, standard modes such as PAL or NTSC are shown shifted. It is not possible to work.

If you use the older P96 drivers from aminet, the problem is solved.

**Comment by midwan** [comment]
@Retro1968 
I don't understand, you mention the P96 driver, but then say that PAL or NTSC screens show the problem? PAL/NTSC screens are not normally related to P96 at all - do you have some kind of mode promotion enabled?

**Comment by Retro1968** [comment]
I can only report what happened in Deluxe Paint V when i choosed there at start a PAL/NTSC resolution and how i solved it. I have nothing other installed. PAL/NTSC Games do not have had that problems.

**Comment by midwan** [comment]
@Retro1968 
It seems to work normally using the latest P96 drivers here (v3.02), both in P96 screens and in PAL/NTSC modes on DPaintV

**Comment by Retro1968** [comment]
I cannot currently test with the P96 3.0.2 because my annual subscription has expired.

I have one more thing to say.
The test is done with amiberry 3.4 beta (2021-02-04) on Manjaro Linux with AmigaOS 3.1.4.1 using as AmiHybrid.

When I activate the virtual mouse driver, activate "Both" and then activate "Magic Mouse untrap", sometimes the amiga mouse pointer AND the Linux mouse pointer appear at the same time even if I do not press the middle mouse button. For example, when leaving a game. The Linux mouse pointer then no longer disappears. Not even when I press the middle mouse button. If I deactivate "Magic Mouse untrap" the problem is solved.

**Comment by midwan** [comment]
@Retro1968 
That is the expected behavior, when using "Show both cursors" - to show both the Linux and the Amiga mouse cursors at the same time, when using the Virtual Mouse Driver.

**Comment by Retro1968** [comment]
So that's the meaning of "Both". I didn't realize that.
Why a virtual mouse driver at all?

**Comment by Retro1968** [comment]
Raspberry Pi 4. Amiberry 3.4 64-Bit (2021-02-07) on Manjaro Linux running Workbench 3.1.4.1.

I ask you people to help testing. I believe that the stuttering described above occurs when "autofire" is activated on port 1 on the joystick.

The jerking does not have to occur immediately. It can arise when you go to the menu with F12 and, for example, execute "Rescan Path", "Update WHDload XML" and "Update Controller DB" and then continue looking around in the menu or working on the workbench again. So try it out for a while. Sometimes quit Amiberry, restart it and then work in it again. Then try out how it behaves when a Linux restart is carried out.

I do not know exactly whether the jerking is related to the activation of Autofire, so please test it.

**Comment by midwan** [comment, marked as answer]
Regarding the virtual mouse driver: enabling it uses absolute mouse positioning, which is helpful if you want to use a tablet, touch controls, lightpen, running Amiberry under VNC or use the "Magic Mouse" release mode (when it goes outside the edges of the window).
If it's not enabled, the emulator uses relative positioning, but then you cannot release the mouse unless you trigger it manually (e.g. with the middle mouse button, if you have that option enabled).

Regarding the testing: Amiberry is being tested daily already by several people, currently the AmiKit beta testers team for example. I'll check if autofire has any effect as you say, but so far nobody else has reported anything like this. Perhaps it's a strange combination of things that we haven't discovered yet...

**Comment by midwan** [comment]
From what you mentioned above:

> But even with Linux 64 bit itself I sometimes have these problems.

This doesn't seem to be an issue in the emulator itself, but an issue on the whole system in your case. Perhaps that's why we can't recreate it anywhere else (at least, not yet).

**Comment by Retro1968** [comment]
As you read above, I tested several Linux distributions. There were always jerking problems under 64 bit, never under 32 bit distributions. I conclude,

1. That my hardware is working properly.
2. Since I have the error with different 64-Bit Linux distributions (Manjaro+Ubuntu), it should be an amiberry error.

If I'm not using amiberry, I can always work with Linux without any problems. So far I had one time a stuttering under Linux after I left amiberry, but I haven't been able to reproduce that error again. Maybe because I use new amiberry 3.4 beta versions very regularly. 

I have therefore tried various amiberry settings for some time. The references to the autofire option grew stronger. However, this is not conclusively certain.

**Comment by Retro1968** [comment]
I suggest doing a Manjaro update with this command before testing:  **sudo pacman -Syu**

If you want to clear the cache of packages that are no longer installed, enter the following command: **sudo pacman -Sc** 

**MAKE A BACKUP COPY OF YOUR SD-CARD BEFORE**

**Reply by midwan** [reply]
I always run the latest updates on my system, so it's definitely not related to that.

**Comment by Retro1968** [comment]
In which mode do you run amiberry as AmiHybrid? Fullscreen, Full-Window or Windowed? Which mode do you recommend Dimitris? In the meantime I have stuttered again, even without autofire.

**Reply by midwan** [reply]
I'm testing all modes, but mostly in Windowed and Full-Window. It's rare that you'd also want to change the resolution (which Fullscreen would do).

**Comment by midwan** [comment]
Moving this to a discussion, as it's going outside the scope of the original issue...

**Comment by Retro1968** [comment]
I completely rebuilt my system with Manjaro Linux 64 Bit, amiberry 3.4 (with new config files) and AmigaOS 3.1.4.1

I currently have the following problem.

If I, for example start on amiberry VLC via Rabbithole, I can return to amiberry with ALT + TAB keys. That's OK. But if I'm on amiberry now, I can't go back to VLC with ALT + TAB; even if I had previously pressed the middle mouse button.

How do I switch between tasks?

**Comment by Retro1968** [comment]
Hello Dimitris,
With the new amiberry 3.4 (2021-02-08) I can switch back to normal between the individual tasks via ALT + TAB. Thank you!

Because of the jerking, I haven't tested long enough after completely reinstalling the system yesterday.

But I can report that there are problems with the "Update Controller DB" option under "Paths".

I had worked without any problems for a while. Then I performed all updates under "Paths". Then I left amiberry and restarted. Then I got error messages, the system booted up anyway, but the workbench resolution was set to NTSC. A restart did not bring any solution. So I recompiled amiberry and everything went back to normal. So for now I will keep my hands off the "Update Controller DB" function.

**Reply by midwan** [reply]
That button just (re)-downloads a file to the conf directory, and I haven't noticed a problem with it so far. It does not (and cannot) mess with anything else in the emulator, and if it fails to download the file for any reason, it should tell you so with a message. Check the logfile as well, it will write details there while downloading.

**Reply by Retro1968** [reply]
> 
> 
> That button just (re)-downloads a file to the conf directory, and I haven't noticed a problem with it so far. It does not (and cannot) mess with anything else in the emulator, and if it fails to download the file for any reason, it should tell you so with a message. Check the logfile as well, it will write details there while downloading.

I just report what happened. I updated, than restarted amiberry.

**Reply by Retro1968** [reply]
I would like to add one more addition to the stuttering.

Today I had very severe stutters on the Linux desktop without having worked with amiberry beforehand. In Manjaro I called "Mouse and Tochpad" in the settings. There were four entries under "Devices". My keyboard was also listed there. After deactivating three devices there, I haven't noticed any more stuttering under Linux and Amiberry.

It seems to have been a Linux problem. But .... I'll test a few more days to be able to say that for good.

**Comment by Retro1968** [comment]
As far as I can tell, I was able to fix the problem. The solution was something else entirely.

I came across it via a detour when I configured the DOS box for the game "Heretic" under RetroPie. I had considerable stutters there.

There are instructions on how to fix the jerking on the Internet.
https://docs.libretro.com/guides/runahead/

However, I have found that I do not need these instructions at all if I use a PS/2 keyboard instead of my existing USB keyboard on the Raspberry Pi 4. Here I recommend the mechanical keyboard "Cherry G80-3000" or "SteelSeries 6Gv2". Although these keyboards also have a USB plug, they are PS/2 keyboards that can be connected to the PC with a USB to PS/2 adapter. 

During my tests I "accidentally" had my "RetroFun! Adapter" with my "VS-7000" joystick unplugged. When it was plugged in again, there were phases of very strong jerking again. But after I had installed the firmware 1.03, the adapter ran perfectly, and without causing any stuttering.

With these two changes, amiberry now also runs smoothly under Manjaro Linux 64 bit ... at least so far. ;-) 

So in the end there were two components that causes troubles. The RetroFun! adapter was probably responsible for the strong stuttering with the old firmware version. The problems have been resolved with the current firmware version 1.03.

In addition, the use of a PS/2 capable keyboard is recommended for use on the Raspberry Pi 4. At least under the DOS box, jerks could be prevented without using a workaround.


Dimitris ... why didn't you tell me that right away. ;-)

Cheers
Retro1968



RetroFun! adapter
https://www.youtube.com/watch?v=j_12IGIzXiQ

VS-7000
https://www.youtube.com/watch?v=5mqFDbie8iw

**Reply by midwan** [reply]
Great that you found a solution, thanks for sharing!
